### PR TITLE
Sync the repository and local store with the disk configuration files and other fixes

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes

This PR changes the implementation of the Repository and Local Zen Store so that they
automatically reload any changes done to the corresponding files stored on disk (e.g.
by an asynchronous CLI command that changes the active stack or registers new stacks).

NOTE: this only works in an initialized ZenML repository and only applies to the stack
configurations (active stack, configured stacks and stack components). Global configuration
changes and changes to the local (in the .zen repository) active profile require some more
refactoring of the relationships between the Repository, GlobalConfiguration and ZenStore
components (see TODO).

Other fixes included in this PR:

* fix the SQL store (currently broken) and update the CI to test it regularly
* fix the Azure step operator (currently broken)
* fix the currently broken migration from legacy .zen repository formats
* make it so that the `zenml stack rename` doesn't need all the integrations installed

Implementation details:

* add a new type of Pydantic Model class, `FileSyncModel`,  that synchronizes the field values
to a designated configuration file stored on the local filesystem automatically, every time an attribute
is read (`__getattribute__`) or set (`__setattr__`).
* change the existing Pydantic classes used to represent the Repository configuration and the Local
Stack Store model to inherit from this new `FileSyncModel` class
* some additional changes were done to remove the code dealing with reading/writing the
Repository configuration and Local Stack Store model from file and rely solely on the functionality
available in the base `FileSyncModel` class

TODO:
- [ ] changing the active profile in the Repository configuration asynchronously is still not automatically reflected in the
currently running ZenML code (switching the active stack works though)
- [ ] same thing with changes in the global configuration (e.g. global active profile and global active stack)

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by htahir1 — APPROVED** [review]
I am completely unfamiliar with this part of the stack so I would say my review is pretty useless right now, unless I spent time understanding it (Which I am happy to do but maybe after the retreat). From the PR notes and a glance over the basics, I would say im happy with this if it works as describes.

I have checked it with changing stacks in a notebook and its now working as expected! But we need to be careful to do a `zenml stack set` rather than just a `zenml stack update`, which doesnt work :-)

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/utils/filesync_model.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    _config_file_timestamp: Optional[float]
+
+    def __init__(self, config_file: str, **kwargs: Any) -> None:
+        """Create a FileSyncModel instance synchronized with a configuration
+        file on disk.
+
+        Args:
+            config_file: configuration file path. If the file exists, the model
+                will be initialized with the values from the file.
+            **kwargs: additional keyword arguments to pass to the Pydantic model
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/utils/filesync_model.py:48]: I don't think I understand the question. They end up being passed to the pydantic super call, albeit overimposed on the values loaded from the configuration file (if exists)

File: src/zenml/utils/filesync_model.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        self._config_file_timestamp = os.path.getmtime(self._config_file)
+
+    def load_config(self) -> None:
+        """Loads the model from the configuration file on disk."""
+        if not fileio.exists(self._config_file):
+            return
+
+        # don't reload the configuration if the file hasn't
+        # been updated since the last load
+        file_timestamp = os.path.getmtime(self._config_file)
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/utils/filesync_model.py:95]: It does, yes :) Thanks for asking though.

File: src/zenml/utils/filesync_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/utils/filesync_model.py:48]: You're absolutely right, I was probably just tired and missed the dict update. I thought they were just not used anywhere, sorry

**Review by schustmi — APPROVED** [review]

**Pull request merged**